lib/puma/request.rb

```diff
@@ -165,9 +165,8 @@ def prepare_response(status, headers, res_body, requests, client)
       # This allows Puma to service connections fairly when the number
       # of concurrent connections exceeds the size of the threadpool.
       force_keep_alive = if @enable_keep_alives
-        requests < @max_fast_inline ||
-        @thread_pool.busy_threads < @max_threads ||
-        !client.listener.to_io.wait_readable(0)
+        client.requests_served < @max_fast_inline ||
```

Comment:
The PR description states:
> Note that all keep-alive connections are closed when the number of requests hits the max_fast_inline setting.

But this logic indicates otherwise. You'll also need the `busy_threads < max_threads` to force the connection to close.

Reply:
I see what you're saying. Previously we had this logic in two places. One of them (here) said "take the keepalive path if any of the following are true" and the other one said "close the connection if requests served is greater than @max_fast_inline."

Now we have only one location that will continue to serve keepalive requests beyond @max_fast_inline as long as not all worker threads are spawned. I think that's okay actually. The @max_fast_inline config was a stop-gap to ensure that one keepalive connection didn't hog too many system resources. If there's still threads we could spawn, there's still capacity in the system. I think instead of adding back a second round of checks, we could add this caveat in the documentation.

In general my recommendation (the one I put on the Heroku docs) suggest setting minimum == maximum and for that case the current docs are already true.

The current docs are wrong anyway after this change:

```ruby
    # The number of requests to attempt inline before sending a client back to
    # the reactor to be subject to normal ordering.
    #
    # The default is 10.
    #
    # @example
    #   max_fast_inline 20
    #
```

So they need to be updated either way. I suggest this:

```ruby
    # A "fast inline" request is a request that is re-using an existing
    # connection via HTTP 1.1 "keepalive". Setting this value will close a connection
    # after the specified number of requests are made when there are no additional
    # Puma threads to spawn. If you want to ensure that a connection is closed after
    # the given number of requests, set minimum thread count equal to maximum thread
    # count.
    #
    # Previous design of Puma would serve these request from the same thread that handled the
    # first request. In that previous model, setting this value would prevent
    # one request from monopolizing too many server resources.
    #
    # Current puma design will serve buffered sequential requests from the same
    # thread but any new requests that come in on the same connection will be
    # sent back to the reactor to read the request body and later be picked up
    # by a thread. With this model resources are distributed more evenly by
    # default.
    #
    # The default is 10.
    #
    # @example
    #   max_fast_inline 1000
    #
```

It's a bit verbose, but I prefer my docs to say more rather than less. FWIW nginx sets this value to 1000 https://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive_requests. 

Eventually: In the long term (not now) we could consider renaming this config and increasing the limit. I'm unsure if the warning nginx gives about memory is intrinsic to the domain or is an implementation detail of how they're managing memory. But if we're able to show that we're not leaking/retaining memory between subsequent repeated requests, then I think raising the limit would be good.